engine/enqueue.go

```diff
@@ -0,0 +1,119 @@
+package engine
+
+import (
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/utils/logging"
+)
+
+type Message struct {
+	OriginID flow.Identifier
+	Payload  interface{}
+}
+
+// MessageStore is the interface to abstract how messages are buffered in memory before
+// being handled by the engine
+type MessageStore interface {
+	Put(*Message) bool
+	Get() (*Message, bool)
+}
+
+type Pattern struct {
+	// Match is a function to match a message to this pattern, typically by payload type.
+	Match MatchFunc
+	// Map is a function to apply to messages before storing them. if not provided, then the message won't get mapped.
+	Map MapFunc
+	// Store is an abstract message store where we will store the message upon receipt.
+	Store MessageStore
+	// OnStore is a hook for functions to be called when a message is stored.
+	OnStore []OnMessageFunc
+}
+
+type OnMessageFunc func(*Message)
+
+type MatchFunc func(*Message) bool
+
+type MapFunc func(*Message) *Message
+
+type MessageHandler struct {
+	log      zerolog.Logger
+	notify   chan<- struct{}
+	patterns []Pattern
+}
+
+func NewMessageHandler(log zerolog.Logger, patterns ...Pattern) (*MessageHandler, <-chan struct{}) {
```

Comment:
Can we have a `struct` as return value in this function? It will allow us to store only one object in caller context instead of two and provides more context than `chan struct{}`

Reply:
I'm not clear about the struct, can you explain what kind of context could be provided with the struct?

Reply:
I mean to have something like
```
type MessageHandler struct {
     handler *MessageHandler
     notifier <-chan struct{}
}

func (h *MessageHandler) Notify() {
       h.notifier <- struct{}{}
}

```

Looks more attractive to me and requires storing only one object instead of 2.

Reply:
I see your point. Yeah makes sense. 

I added a method `GetNotifier` to the MessageHandler, so that the Publisher can publish message with `handler.Process`, and the Consumer can consume messages with `<- handler.GetNotifier()`. 

So now, the compliance engine doesn't need to store the `notifier` channel, since the handler will store it anyway